nemo/collections/tts/models/fastpitch_hifigan_e2e.py

```diff
@@ -0,0 +1,415 @@
+            total_loss += pitch_loss
+            # loss_feat_mp = self.feat_matching_loss(real_feat_mp, gen_feat_mp)
+            # loss_feat_ms = self.feat_matching_loss(real_feat_ms, gen_feat_ms)
+            # total_loss += loss_feat_mp + loss_feat_ms
+            # self.log(name="loss_gen_disc_feat", value=loss_feat_mp + loss_feat_ms)
+            # self.log(name="loss_gen_disc_feat_ms", value=loss_feat_ms)
+            # self.log(name="loss_gen_disc_feat_mp", value=loss_feat_mp)
+
+            self.log(name="loss_gen_mel", value=loss_mel)
+            self.log(name="loss_gen_disc", value=loss_gen_mp + loss_gen_ms)
+            self.log(name="loss_gen_disc_mp", value=loss_gen_mp)
+            self.log(name="loss_gen_disc_ms", value=loss_gen_ms)
+            self.log(name="loss_gen_duration", value=dur_loss)
+            self.log(name="loss_gen_pitch", value=pitch_loss)
+
+            # Log images to tensorboard
+            if self.log_train_images:
+                self.log_train_images = False
+
+                if self.logger is not None and self.logger.experiment is not None:
+                    self.tb_logger.add_image(
+                        "train_mel_target",
+                        plot_spectrogram_to_numpy(real_spliced_spec[0].data.cpu().numpy()),
+                        self.global_step,
+                        dataformats="HWC",
+                    )
+                    spec_predict = pred_spliced_spec[0].data.cpu().numpy()
+                    self.tb_logger.add_image(
+                        "train_mel_predicted",
+                        plot_spectrogram_to_numpy(spec_predict),
+                        self.global_step,
+                        dataformats="HWC",
+                    )
+            self.log(name="loss_gen", prog_bar=True, value=total_loss)
+            return total_loss
+
+    def validation_step(self, batch, batch_idx):
+        audio, audio_lens, text, text_lens, durs, pitch, speakers = batch
+        mels, mel_lens = self.preprocessor(audio, audio_lens)
+
+        audio_pred, _, log_durs_predicted, _ = self(text=text, durs=None, pitch=None, splice=False)
+        audio_length = torch.sum(torch.clamp(torch.exp(log_durs_predicted - 1), 0), axis=1)
+        audio_pred.squeeze_()
+        pred_spec, _ = self.melspec_fn(audio_pred, audio_length)
+        loss = self.mel_val_loss(
+            spec_pred=pred_spec, spec_target=mels, spec_target_len=mel_lens, pad_value=-11.52, transpose=False
+        )
+
+        return {
+            "val_loss": loss,
+            "audio_target": audio if batch_idx == 0 else None,
+            "audio_pred": audio_pred.squeeze() if batch_idx == 0 else None,
+        }
+
+    def validation_epoch_end(self, outputs):
+        if self.tb_logger is not None:
+            _, audio_target, audio_predict = outputs[0].values()
+            if not self.logged_real_samples:
+                self.tb_logger.add_audio("val_target", audio_target[0].data.cpu(), self.global_step, 22050)
```

Comment:
Should the sample rate come from the config or data loader?

Reply:
This still shows as hard-coded 22050 from my end, can you double-check?

Reply:
Should be fixed now